src/object.cr

```diff
@@ -86,72 +86,65 @@ class Object
     hash(Crystal::Hasher.new).result
   end
 
-  # Returns a string representation of this object.
+  # Returns a nicely readable and concise string representation of this object,
+  # typically intended for users.
   #
-  # Descendants must usually **not** override this method. Instead,
-  # they must override `to_s(io)`, which must append to the given
-  # IO object.
+  # This method should usually **not** be overridden. It delegates to
+  # `#to_s(IO)` which can be overridden for custom implementations.
+  #
+  # See `#inspect` for related method.
   def to_s : String
     String.build do |io|
       to_s io
     end
   end
 
-  # Appends a `String` representation of this object
-  # to the given `IO` object.
-  #
-  # An object must never append itself to the io argument,
-  # as this will in turn call `to_s(io)` on it.
-  abstract def to_s(io : IO) : Nil
-
-  # Returns a `String` representation of this object suitable
-  # to be embedded inside other expressions, sometimes providing
-  # more information about this object.
+  # Prints a nicely readable and concise string representation of this object,
+  # typically intended for users, to *io*.
   #
-  # `#inspect` (and `#inspect(io)`) are the methods used when
-  # you invoke `#to_s` or `#inspect` on an object that holds
-  # other objects and wants to show them. For example when you
-  # invoke `Array#to_s`, `#inspect` will be invoked on each element:
+  # This method is called when an object is used in string interpolation:
+  # In `"foo #{bar} baz"`, the value interpolated for `#{foo}` is the result of
+  # `foo.to_s`.
   #
-  # ```
-  # ary = ["one", "two", "three, etc."]
-  # ary.inspect # => ["one", "two", "three, etc."]
-  # ```
+  # Implementations must not append `self` to *io* which would lead to an
```

Comment:
I thin is clearer to say that self.to_s should not be called since that method will call this one.

Is that what you are trying to point, right?

Reply:
I'd assume the connection to `#to_s` without IO should be relatively obvious. But this is about `IO#<<`:
```cr
def to_s(io)
  io << self
end
```
Maybe add `io << self` as an example?

Reply:
Maybe we need to say that when evaluating expressions like `"foo #{bar} baz"` or `io << bar`, the `bar.to_s(io)` is called.

Note that the interpolation example above uses `"foo #{bar} baz"` but then mention `#{foo}` and `foo.to_s`

Reply:
I've updated the documentation, it should be clearer now.